---
title: "Case study - deconvolution"
author:
  - Maria Doyle, Peter MacCallum Cancer Centre^[maria.doyle@petermac.org]
  - Stefano Mangiola, Walter and Eliza Hall Institute^[mangiola.s@wehi.edu.au]
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Case study - deconvolution}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)

# bibliography: "`r file.path(system.file(package='tidytranscriptomics', 'vignettes'), 'workshop_1_hour.bib')`"

```

Recommended Background Reading
[Introduction to R for Biologists](https://melbournebioinformatics.github.io/r-intro-biologists/intro_r_biologists.html)

```{r, echo=FALSE, out.width = "200px"}
knitr::include_graphics("../inst/vignettes/tidybulk_logo.png")
```

```{r}

library(tidyverse)
library(tidybulk)
library(survival)
```
```{r echo = FALSE}
# Use colourblind-friendly colours
friendly_cols <- dittoSeq::dittoColors()

# Set theme
custom_theme <-
  list(
    scale_fill_manual(values = friendly_cols),
    scale_color_manual(values = friendly_cols),
    theme_bw() +
      theme(
        panel.border = element_blank(),
        axis.line = element_line(),
        panel.grid.major = element_line(size = 0.2),
        panel.grid.minor = element_line(size = 0.1),
        text = element_text(size = 6),
        legend.position = "bottom",
        #aspect.ratio = 1,
        strip.background = element_blank(),
        axis.title.x = element_text(margin = margin(t = 10, r = 10, b = 10, l = 10)),
        axis.title.y = element_text(margin = margin(t = 10, r = 10, b = 10, l = 10)),
        axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1)
      )
  )
```


## Introduction

With `tidybulk` it is easy to infer tissue composition and perform hypothesis testing on cell type proportion; analogously to testing differential abundance.

### Deconvolution

For this example we will use a subset of the breast cancer TCGA dataset. 

```{r}
BRCA_tidy = 
	tidytranscriptomics::BRCA %>%
	tidybulk(patient, transcript, count)

BRCA_tidy
```

We can easily infer tissue cellularity across a range of methods (Cibersort, EPIC and ls_fit) with LM22 or custom signature

```{r}
BRCA_cell_type = 
	BRCA_tidy %>%
	deconvolve_cellularity(action="get")

BRCA_cell_type
```

Cell type proportions are added to the tibble as new columns. The prefix makes it easy to reshape the data frame if needed, for visualisation or further analyses.

```{r}
BRCA_cell_type_long = 
	BRCA_cell_type %>%
	
	# Reshape
	pivot_longer(
		contains("cibersort"), 
		names_prefix = "cibersort: ", 
		names_to = "cell_type",
		values_to = "proportion"
	) 

BRCA_cell_type_long

```

Note: Variable creation is functional to education, however is not needed nor recommended in normal circumstances.

```{r out.width = "100%", dpi=300}
BRCA_cell_type_long %>%
	
	# Plot
  ggplot(aes(x=patient, y=proportion, fill=cell_type)) +
  geom_bar(stat = "identity") +
  custom_theme +
  theme(
    legend.key.size = unit(1, "mm"),
    legend.text = element_text(size = 4)
  )
  

```

We can visualise the tissue composition similarity performing a dimensionality reduction on cell type and proportion (rather than on transcript and counts)

```{r out.width = "70%", dpi=300}

BRCA_cell_type_long %>%
	
	# Add some noise to the proportion inference to avoid PCA error
	# A better (but more complex) way would be to eliminate the zero variance cell types 
	mutate(proportion = proportion + rnorm(n(), 0, 1e-10) %>% abs) %>%
	
	reduce_dimensions(
		patient, 
		cell_type, 
		proportion, 
		method="PCA", 
		action="get"
	) %>%
	
	ggplot(aes(PC1, PC2, label=patient)) +
	geom_point(color="red") +
	ggrepel::geom_text_repel(size=2) +
	custom_theme

```

### Hypothesis testing - survival analyses

We can also perform differential tissue composition analyses, similarly to how we performed differential transcript abundance analyses. We can perform our analyses on censored (survival analyses) or known factor of interest.

```{r}
library(survival)

BRCA_tidy_survival = 
	BRCA_tidy %>%
	test_differential_cellularity(Surv(PFI.time.2, PFI.2) ~ .	) %>%
	arrange(p.value)

BRCA_tidy_survival %>%
	dplyr::select(.cell_type, p.value, everything())
```

We can visualise the cell type proportions for the most associated cell types

```{r out.width = "70%", dpi=300}
BRCA_tidy_survival %>%
	slice(1:2) %>%
	unnest(cell_type_proportions) %>%
	ggplot(aes(PFI.time.2, .proportion, color = factor(PFI.2))) +
	geom_point() +
	facet_wrap(~ .cell_type) +
	scale_x_log10() +
	scale_y_continuous(trans = "logit") +
	custom_theme
```



## Contributing
If you want to suggest improvements for this workshop or ask questions, you can do so as described [here](https://github.com/stemangiola/tidytranscriptomics/blob/master/CONTRIBUTING.md).

## Reproducibility
Record package and version information with `sessionInfo`

```{r}
sessionInfo()
```

## References
