---
title: "Case study - single cell SingleCellExperiment"
author:
  - Maria Doyle, Peter MacCallum Cancer Centre^[maria.doyle@petermac.org]
  - Stefano Mangiola, Walter and Eliza Hall Institute^[mangiola.s@wehi.edu.au]
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Case study - single cell SingleCellExperiment}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval=FALSE # temporarily not evaluating until get website to update and sort out what datasets to use
)

# bibliography: "`r file.path(system.file(package='tidytranscriptomics', 'vignettes'), 'workshop_1_hour.bib')`"

```

Recommended Background Reading
[Introduction to R for Biologists](https://melbournebioinformatics.github.io/r-intro-biologists/intro_r_biologists.html)

```{r, echo=FALSE, out.width = "200px"}
knitr::include_graphics("../inst/vignettes/tidybulk_logo.png")
```

SingleCellExperiment is a very popular container for single cell RNA sequencing data [@sce].

tidySCE provides a bridge between the SingleCellExperiment single-cell package [@sce] and the tidyverse [@wickham2019welcome]. It enables the display of the
SingleCellExperiment object as a tidyverse tibble, and provides SingleCellExperiment-compatible *dplyr*, *tidyr*, *ggplot* and *plotly* functions.


```{r message=FALSE, warning=FALSE}
# load additional libraries
library(biocasia2020tidytranscriptomics)
library(dplyr)
library(purrr)
library(stringr)
library(SummarizedExperiment)
library(SingleCellExperiment)
library(scater)
library(scran)
library(igraph)
library(SingleR)
library(tidyHeatmap)
library(tidySCE)
```

```{r echo = FALSE}
# Use colourblind-friendly colours
friendly_cols <- dittoSeq::dittoColors()

# Set theme
custom_theme <-
  list(
    scale_fill_manual(values = friendly_cols),
    scale_color_manual(values = friendly_cols),
    theme_bw() +
      theme(
        panel.border = element_blank(),
        axis.line = element_line(),
        panel.grid.major = element_line(size = 0.2),
        panel.grid.minor = element_line(size = 0.1),
        text = element_text(size = 6),
        legend.position = "bottom",
        #aspect.ratio = 1,
        strip.background = element_blank(),
        axis.title.x = element_text(margin = margin(t = 10, r = 10, b = 10, l = 10)),
        axis.title.y = element_text(margin = margin(t = 10, r = 10, b = 10, l = 10)),
        axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1)
      )
  )
```

## Create `tidySCE`

This is a SingleCellExperiment object but it is evaluated as tibble. So it is fully compatible both with SingleCellExperiment and tidyverse APIs. 

```{r}
pbmc_small_tidy <- biocasia2020tidytranscriptomics::pbmc_small %>% tidy()
```

**It looks like a tibble**

```{r}
pbmc_small_tidy
```

**But it is a SingleCellExperiment object after all**

```{r}
assayNames(pbmc_small_tidy) # from SummarizedExperiment
```

## Polish the data

We can interact with our object as we do with any tibble. In this case we want to polish an annotation column.

```{r}
pbmc_small_polished <-
	pbmc_small_tidy %>%
	
	# Clean groups
	mutate(groups = groups %>% str_remove("^g")) %>%
	
	# Extract sample
	extract(file, "sample", "../data/sample([a-z0-9]+)/outs.+")
pbmc_small_polished
```

## Calculate the log of the scaled counts

We can treat `pbmc_small_polished` as a [*SingleCellExperiment* object](https://www.bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html)

```{r}
counts <- assay(pbmc_small_polished, "counts")                         # from SummarizedExperiment
libsizes <- colSums(counts)
size.factors <- libsizes/mean(libsizes)
logcounts(pbmc_small_polished) <- log2(t(t(counts)/size.factors) + 1)  # from SingleCellExperiment
assayNames(pbmc_small_polished)                                        # from SummarizedExperiment
```

## Exploratory analyses

Here we plot abundance of two transcripts for each group.

```{r}
pbmc_small_polished_abundance = 
  pbmc_small_polished %>%
  
  # Extract abundance
  join_transcripts(transcripts=c("HLA-DRA", "TCL1A")) 
  
pbmc_small_polished_abundance
```

We proceed to plot. 
_Note: we don't need to create many of these temporary variables. They are created here for educational purposes only._

```{r}
pbmc_small_polished_abundance %>%
  ggplot2::ggplot(aes(groups, abundance_counts + 1, fill=groups)) +
  geom_boxplot(outlier.shape=NA) +
  geom_jitter(alpha=0.5, width=0.2) +
  scale_y_log10() +
  facet_wrap(~transcript, scales="free_y") +
  custom_theme
```

## Reduce dimensions

### PCA

We proceed with data processing with Bioconductor packages, such as *scran* [@lun2016pooling] and *scater* [@mccarthy2017scater].

```{r preprocess}
variable_genes <-
    pbmc_small_polished %>%
    modelGeneVar() %>%                 # from scran
    getTopHVGs(prop=0.1)               # from scran
# Perform PCA with scater
pbmc_small_pca <-
    pbmc_small_polished %>%
    runPCA(subset_row=variable_genes)  # from scater
pbmc_small_pca
```

If a tidyverse-compatible package is not included in the tidySCE collection, we can use `as_tibble` to permanently convert `tidySCE` into a tibble.

```{r pc_plot, message=FALSE}
# Create pairs plot with GGally
pbmc_small_pca %>%
    as_tibble() %>%
    GGally::ggpairs(columns=4:8, ggplot2::aes(colour=groups)) +
    custom_theme
```


### UMAP

Beside PCA which is a linear dimensionality reduction, we can apply neighbour aware methods such as UMAP, to better define locally similar cells. We can calculate the first 3 UMAP dimensions using the scater framework.

```{r umap}
pbmc_small_UMAP <-
    pbmc_small_pca %>%
    runUMAP(ncomponents=3)           # from scater
pbmc_small_UMAP
```

## Identify clusters 

We can proceed with cluster identification with *scran*.

```{r cluster}
pbmc_small_cluster <- pbmc_small_UMAP
# Assign clusters to the 'colLabels' 
# of the SingleCellExperiment object
colLabels(pbmc_small_cluster) <-                                          # from SingleCellExperiment
    pbmc_small_pca %>%
    buildSNNGraph(use.dimred="PCA") %>%                                   # from scran - shared nearest neighbor
    cluster_walktrap() %$%                                                # from igraph
    membership %>%
    as.factor()
# Reorder columns
pbmc_small_cluster %>% select(label, everything())
```

Now we can interrogate the object as if it was a regular tibble data frame.

```{r cluster count}
pbmc_small_cluster %>% count(groups, label)
```

And we can plot them using 3D plot using plotly.

```{r umap plot, eval=FALSE}
pbmc_small_cluster %>%
  plot_ly(
    x = ~UMAP1,
    y = ~UMAP2,
    z = ~UMAP3,
    color = ~ label,
    colors = friendly_cols[1:5]
  )
```


```{r, echo=FALSE}
knitr::include_graphics("../inst/vignettes/plotly.png")
```

## Plotting heatmap 

### Using scater

```{r}
plotHeatmap(                                  # from scater
  pbmc_small_cluster, 
  features=variable_genes, 
  columns=order(pbmc_small_cluster$label), 
  colour_columns_by=c("label")
) 
```

### Using tidyverse

```{r}
pbmc_small_cluster %>%
  
  # Get transcript abundance
  join_transcripts(transcripts=variable_genes) %>%
  
  # Plot heatmap
  heatmap(                                            # from tidyHeatmap
    .row = transcript,
    .column = cell, 
    .value = abundance_logcounts
  ) %>%
  
  # Add annotation
  add_tile(label) %>%
  add_point(PC1)
```

## Cell type classification

We can infer cell type identities using *SingleR* [@aran2019reference] and manipulate the output using tidyverse. SingleR [accepts any log-normalised transcript abundance matrix.](https://bioconductor.org/packages/devel/bioc/vignettes/SingleR/inst/doc/SingleR.html)

```{r eval = FALSE}
# DO NOT EXECUTE - CELLDEX DEPENDENCY NOT IN BIOCONDUCTOR YET
# Get cell type reference data
hpca <- HumanPrimaryCellAtlasData()
# Infer cell identities
cell_type_df <-
	
  # extracting counts from SingleCellExperiment object
  assays(pbmc_small_cluster)$logcounts%>%
	# SingleR
  SingleR(
    ref = hpca,
    labels = hpca$label.main,
    method = "cluster",
    clusters = pbmc_small_cluster %>% pull(label)
  ) %>%
	
	# Formatting results
  as.data.frame() %>%
  as_tibble(rownames = "label") %>%
  select(label, first.labels)
```

We have pre-calculated the cluster classification.

```{r}
cell_type_df <- biocasia2020tidytranscriptomics::cell_type_df
cell_type_df
```

```{r }
# Join UMAP and cell type info
pbmc_small_cell_type <-
  pbmc_small_cluster %>%
  left_join(
    cell_type_df, 
  	by = "label"
  )
pbmc_small_cell_type 
```

We can easily summarise the results. For example, we can see how cell type classification overlaps with cluster classification.

```{r }
pbmc_small_cell_type %>%
  count(label, first.labels)
```


##  Nested analyses

A powerful tool we can use with tidySCE is `nest`. We can easily perform independent analyses on subsets of the dataset. First we classify cell types in lymphoid and myeloid; then, nest based on the new classification

```{r}
pbmc_small_nested <-
  pbmc_small_cell_type %>%
  filter(first.labels != "Platelets") %>%
  mutate(cell_class = 
           if_else(
             `first.labels` %in% c("Macrophage", "Monocyte"),
             "myeloid", 
             "lymphoid"
           )
          ) %>%
  nest(data = -cell_class)
pbmc_small_nested
```

Now we can independently perform analyses for the lymphoid and myeloid cell subsets.

```{r, warning=FALSE}
pbmc_small_nested %>%
mutate(variable_genes = map_chr(
  data, ~ .x %>%
    modelGeneVar() %>%
    getTopHVGs(prop=0.05) %>% 
    paste(collapse=", ")
)) 
```

## Key Points
- Some basic steps of a single-cell RNA sequencing analysis are dimensionality reduction, cluster identification and cell type classification
- `tidySCE` is an invisible layer that operates on a `SingleCellExperiment` object and enables us to visualise and manipulate data as if it were a tidy data frame.
- `tidySCE` object is a `SingleCellExperiment object` so it can be used with any `SingleCellExperiment` compatible method


```{poll class.source="poll"}
Poll: What would be your preferred name for the package currently called tidySCE?
```


# Contributing
If you want to suggest improvements for this workshop or ask questions, you can do so as described [here](https://github.com/stemangiola/biocasia2020_tidytranscriptomics/blob/master/CONTRIBUTING.md).

# Reproducibility
Record package and version information with `sessionInfo`

```{r}
sessionInfo()
```

# References